### PR TITLE
Adding-emitter example for User Guide

### PR DESCRIPTION
**Pull request by mwichmann**
Add an example on adding an emitter to an existing builder, based on a mailing-list discussion with Jeremy Elson.

Also:

* Try to model "best practice" by changing instances of hand-modifying `env['ENV']['PATH']` in examples to using `AppendENVPath` function.
* Fixed a couple of Py2-style print statements.
* Tweaked markup on a couple of instances of site_scons.
* Black-formatted examples in the two userguide sections that were otherwise modified.
* Dropped old cons doc content which was included as a big   comment at the fromt of builders-writing.xml.

No scons code changes in this PR.

Signed-off-by: Mats Wichmann <mats@linux.com>

## Contributor Checklist:

* [ ] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [X] I have updated `CHANGES.txt` (and read the `README.rst`)
* [X] I have updated the appropriate documentation

### CODE REVIEW

**Comment by mwichmann** [comment]
@jelson this is intended to describe the situation you found... it's little awkward to dig through the xml here, the only part of interest is about line 814 in  `doc/user/builders-writing.xml` - did I captuer this in a somewhat understandable way?

**Comment by jelson** [comment]
> @jelson this is intended to describe the situation you found... it's little awkward to dig through the xml here, the only part of interest is about line 814 in `doc/user/builders-writing.xml` - did I captuer this in a somewhat understandable way?

yes, I think this would have been more "self-service" than writing to the mail list for help. Thanks for improving the docs!

File: doc/user/builders-writing.xml [hunk 1]
```diff
@@ -43,78 +43,6 @@
   OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
   WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
--->
```
Comment (bdbaddog) [inline doc/user/builders-writing.xml:46]: I think worth leaving this in commented out for now.
It's been copied from cons and may have value.. if we get a chance to implement?

File: doc/user/builders-writing.xml [hunk 2]
```diff
@@ -188,9 +116,8 @@ bld = Builder(action = 'foobuild &lt; $SOURCE &gt; $TARGET')
     <scons_example name="builderswriting_ex1">
        <file name="SConstruct">
 bld = Builder(action = 'foobuild &lt; $SOURCE &gt; $TARGET')
-env = Environment(BUILDERS = {'Foo' : bld})
-import os
-env['ENV']['PATH'] = env['ENV']['PATH'] + os.pathsep + os.getcwd()
+env = Environment(BUILDERS={'Foo': bld})
+env.AppendENVPath('PATH', os.getcwd())
```
Comment (bdbaddog) [inline doc/user/builders-writing.xml:121]: Don't we need import os here?

Comment (mwichmann) [reply doc/user/builders-writing.xml:121]: not sure how it got dropped - or why the example builds without it.

Comment (mwichmann) [reply doc/user/builders-writing.xml:121]: I checked all the examples, found a couple of other places with the same issue, and some now-unused example files to removed.

**Comment by mwichmann** [comment]
Hmmm, did github change to start honoring [ci skip] now?

File: doc/user/builders-writing.xml (hunk 1 above)
Comment (mwichmann) [reply doc/user/builders-writing.xml:46]: > I think worth leaving this in commented out for now.
> It's been copied from cons and may have value.. if we get a chance to implement?

I'll of course put it back if you want, it's your call, and costs basically nothing to do.  But I think this kind of wishlist-item works better stored somewhere else. It's been 20 years, and at current contribution levels implementation seems... unlikely. I'm hereby expressing my preference for the docs to be cleaner, not cluttered with "what if someday?". Wiki perhaps?

**Comment by bdbaddog** [comment]
> Hmmm, did github change to start honoring [ci skip] now?

Hmm. strange, it looks that way.

File: doc/user/builders-writing.xml (hunk 1 above)
Comment (bdbaddog) [reply doc/user/builders-writing.xml:46]: Maybe move to github issue as enhancement?
That way it's not lost, but doesn't clutter up the actual docs.

**Comment by mwichmann** [comment]
Not like they'd want to tell us or anything...

File: doc/user/builders-writing.xml (hunk 1 above)
Comment (mwichmann) [reply doc/user/builders-writing.xml:46]: I'm fine with that, as long as you don't worry about issues being cluttered with inactive stuff (have noticed this only seems to bother the maintainer with PRs, not issues :) )

Comment (bdbaddog) [reply doc/user/builders-writing.xml:46]: I'm fine with having many enhancements filed, we can add to projects like "wishlist 1".. or such.
It's really the # of non-enhancement issues (aka bugs) IMHO which are more problematic.

Comment (mwichmann) [reply doc/user/builders-writing.xml:46]: Parked in #3887.

**Comment by mwichmann** [comment]
> Hmmm, did github change to start honoring [ci skip] now?

Yes, this is confirmed now, about a week ago it started working. https://github.blog/changelog/2021-02-08-github-actions-skip-pull-request-and-push-workflows-with-skip-ci/

File: doc/user/builders-writing.xml [hunk 3]
```diff
@@ -187,10 +115,10 @@ bld = Builder(action = 'foobuild &lt; $SOURCE &gt; $TARGET')
 
     <scons_example name="builderswriting_ex1">
        <file name="SConstruct">
-bld = Builder(action = 'foobuild &lt; $SOURCE &gt; $TARGET')
-env = Environment(BUILDERS = {'Foo' : bld})
 import os
-env['ENV']['PATH'] = env['ENV']['PATH'] + os.pathsep + os.getcwd()
+bld=Builder(action = 'foobuild &lt; $SOURCE &gt; $TARGET')
+env = Environment(BUILDERS={'Foo': bld})
+env.AppendENVPath('PATH', os.getcwd())
```
Comment (bdbaddog) [inline doc/user/builders-writing.xml:121]: Would Dir('.') work here? instead of os.getcwd()?

Comment (mwichmann) [reply doc/user/builders-writing.xml:121]: Made this change, after checking it indeed works.

File: doc/user/builders-writing.xml [hunk 4]
```diff
@@ -863,7 +816,92 @@ cat
...
+
+    <scons_example name="builders_adding_emitter">
+
+      <file name="SConstruct" printme="1">
+def map_emitter(target, source, env):
+    target.append(map_filename)
+    return target, source
+
+env = Environment()
+map_filename = "${TARGET.name}" + ".map"
```
Comment (bdbaddog) [inline doc/user/builders-writing.xml]: Is this gonna work? it's referenced in line 868.

File: doc/user/builders-writing.xml (hunk 3 above)
Comment (mwichmann) [reply doc/user/builders-writing.xml:121]: Made that change a few places.

**Pull request merged**